File: app/src/main/java/com/example/purepawapp/ui/cart/CartFragment.java
package com.example.purepawapp.ui.cart;

import android.os.Bundle;
import android.view.View;

import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.purepawapp.R;
import com.example.purepawapp.data.repository.CartRepository;
import com.example.purepawapp.data.repository.RepoCallback;
import com.example.purepawapp.databinding.FragmentCartBinding;
import com.example.purepawapp.di.ServiceLocator;
import com.example.purepawapp.ui.common.BaseFragment;
import com.example.purepawapp.util.CheckoutConfig;
import com.example.purepawapp.util.CurrencyUtils;
import com.example.purepawapp.util.ViewUtils;

public class CartFragment extends BaseFragment<FragmentCartBinding> {

    private CartRepository cartRepository;

    private final CartAdapter adapter = new CartAdapter(
            item -> ServiceLocator.getCartRepository().updateQuantity(item.getProduct().getId(), item.getVariant().getId(), item.getQuantity() + 1),
            item -> ServiceLocator.getCartRepository().updateQuantity(item.getProduct().getId(), item.getVariant().getId(), item.getQuantity() - 1),
            item -> ServiceLocator.getCartRepository().removeFromCart(item.getProduct().getId(), item.getVariant().getId())
    );

    public CartFragment() {
        super(FragmentCartBinding::inflate);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        cartRepository = ServiceLocator.getCartRepository();

        getBinding().rvCartItems.setLayoutManager(new LinearLayoutManager(requireContext()));
        getBinding().rvCartItems.setAdapter(adapter);

        getBinding().btnApplyPromo.setOnClickListener(v -> {
            String code = getBinding().etPromoCode.getText() != null ? getBinding().etPromoCode.getText().toString().trim() : "";
            if (code.isEmpty()) {
                ViewUtils.toast(this, "Vui lòng nhập mã giảm giá");
                return;
            }
            showLoading();
            cartRepository.applyPromoCode(code, new RepoCallback<>() {
                @Override
                public void onSuccess(com.example.purepawapp.data.model.Promotion result) {
                    ViewUtils.toast(CartFragment.this, "Áp dụng mã giảm giá thành công");
                    hideLoading();
                }

                @Override
                public void onError(Exception error) {
                    ViewUtils.toast(CartFragment.this, error.getMessage() != null ? error.getMessage() : "Mã giảm giá không hợp lệ");
                    hideLoading();
                }
            });
        });

        getBinding().btnCheckout.setOnClickListener(v -> {
            if (cartRepository.getItems().isEmpty()) {
                ViewUtils.toast(this, "Giỏ hàng của bạn đang trống");
                return;
            }
            NavHostFragment.findNavController(this).navigate(R.id.action_cartFragment_to_checkoutAddressFragment);
        });

        cartRepository.getItemsLiveData().observe(getViewLifecycleOwner(), items -> renderCart());
        cartRepository.getAppliedPromotionLiveData().observe(getViewLifecycleOwner(), promo -> renderCart());
    }

    private void renderCart() {
        var items = cartRepository.getItems();
        adapter.submitList(items);

        boolean isEmpty = items.isEmpty();
        getBinding().emptyState.setVisibility(isEmpty ? View.VISIBLE : View.GONE);
        getBinding().cartContent.setVisibility(isEmpty ? View.GONE : View.VISIBLE);

        double subtotal = cartRepository.subtotal();
        double discount = cartRepository.discount();
        double shipping = isEmpty ? 0.0 : CheckoutConfig.SHIPPING_FEE;
        double total = cartRepository.total() + shipping;

        getBinding().tvSubtotal.setText(CurrencyUtils.toVndString(subtotal));
        getBinding().tvShipping.setText(CurrencyUtils.toVndString(shipping));
        getBinding().tvDiscount.setText("-" + CurrencyUtils.toVndString(discount));
        getBinding().tvTotal.setText(CurrencyUtils.toVndString(total));
        getBinding().tvTotalBottom.setText(CurrencyUtils.toVndString(total));
        getBinding().btnCheckout.setEnabled(!isEmpty);
    }
}

File: app/src/main/java/com/example/purepawapp/ui/cart/CartFragment.kt
package com.example.purepawapp.ui.cart

import android.os.Bundle
import android.view.View
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.purepawapp.R
import com.example.purepawapp.databinding.FragmentCartBinding
import com.example.purepawapp.di.ServiceLocator
import com.example.purepawapp.ui.common.BaseFragment
import com.example.purepawapp.util.toVndString
import com.example.purepawapp.util.toast
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.launch

private const val SHIPPING_FEE = 25000.0

class CartFragment : BaseFragment<FragmentCartBinding>(FragmentCartBinding::inflate) {

    private val cartRepository get() = ServiceLocator.cartRepository

    private val adapter = CartAdapter(
        onIncrease = { item -> cartRepository.updateQuantity(item.product.id, item.variant.id, item.quantity + 1) },
        onDecrease = { item -> cartRepository.updateQuantity(item.product.id, item.variant.id, item.quantity - 1) },
        onRemove = { item -> cartRepository.removeFromCart(item.product.id, item.variant.id) }
    )

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.rvCartItems.layoutManager = LinearLayoutManager(requireContext())
        binding.rvCartItems.adapter = adapter

        binding.btnApplyPromo.setOnClickListener {
            val code = binding.etPromoCode.text?.toString().orEmpty().trim()
            if (code.isEmpty()) {
                toast("Vui lòng nhập mã giảm giá")
                return@setOnClickListener
            }
            viewLifecycleOwner.lifecycleScope.launch {
                cartRepository.applyPromoCode(code)
                    .onSuccess { toast("Áp dụng mã giảm giá thành công") }
                    .onFailure { toast(it.message ?: "Mã giảm giá không hợp lệ") }
            }
        }

        binding.btnCheckout.setOnClickListener {
            if (cartRepository.items.value.isEmpty()) {
                toast("Giỏ hàng của bạn đang trống")
                return@setOnClickListener
            }
            findNavController().navigate(R.id.action_cartFragment_to_checkoutAddressFragment)
        }

        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                combine(cartRepository.items, cartRepository.appliedPromotion) { items, _ -> items }
                    .collect { items ->
                        adapter.submitList(items)

                        val isEmpty = items.isEmpty()
                        binding.emptyState.visibility = if (isEmpty) View.VISIBLE else View.GONE
                        binding.cartContent.visibility = if (isEmpty) View.GONE else View.VISIBLE

                        val subtotal = cartRepository.subtotal()
                        val discount = cartRepository.discount()
                        val shipping = if (isEmpty) 0.0 else SHIPPING_FEE
                        val total = cartRepository.total() + shipping

                        binding.tvSubtotal.text = subtotal.toVndString()
                        binding.tvShipping.text = shipping.toVndString()
                        binding.tvDiscount.text = "-${discount.toVndString()}"
                        binding.tvTotal.text = total.toVndString()
                        binding.tvTotalBottom.text = total.toVndString()
                        binding.btnCheckout.isEnabled = !isEmpty
                    }
            }
        }
    }
}
